Support tables with multiple header rows

SandraFer opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Some guidelines are provided like this, in HTML comments, to expedite the code review before merging your contribution.

First, please make sure that you have read the documentation page on fpdf2 development:
https://py-pdf.github.io/fpdf2/Development.html

If you're new to contributing to open source projects,
you might find this free video course helpful: http://kcd.im/pull-request
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
Fixes #938 

This PR implements support for tables with multiple heading rows by adding a new argument `num_heading_rows` to the `Table` constructor with default value = 1. Since this argument can conflict with the existing `first_row_as_headings`, the following applies
- if `num_heading_rows==1`: The old `first_row_as_headings` is respected. This means that if `first_row_as_headings` is `False`, the private attribute `_num_heading_rows` is set to 0.
- for any other value of `num_heading_rows`, we raise an error in case the passed value of `first_row_as_headings` is in conflict (so, error is raised if `first_row_as_headings is True and num_heading_rows==0` or `first_row_as_headings is False and num_heading_rows>1`

Other changes worth mentioning here;
- For the `TableBordersLayout.NO_HORIZONTAL_LINES` border layout, horizontal lines between headings rows are drawn.
- A unit test for a table with two header rows and a page break is added.

<!-- Have you done all of these things?  -->
**Checklist**:

<!-- To check an item, place an "x" in the box like so: "- [x] Item description"
     Add "N/A" to the end of each line that's irrelevant to your changes -->

- [ ] The GitHub pipeline is OK (green), <!-- The maintainers will trigger it if this is your 1st contribution -->
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged <!-- In your opinion, can this be merged as soon as it's reviewed? Else, this can be turned into a Draft PR -->

- [x] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [x] A mention of the change is present in `CHANGELOG.md`

<!-- Feel free to add additional comments, and to ask questions if some of those guidelines are unclear to you! -->

<!--
Once a PR is merged, maintainers will add your name to the contributors table in README.md.
If they forget, or you do not wish this to happen, please mention it.
-->

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/py-pdf/fpdf2/blob/master/LICENSE).

SandraFer commented:
@Lucas-C you had volunteered to review this but I can't seem to change the assigned reviewer?

andersonhc commented:
> @Lucas-C you had volunteered to review this but I can't seem to change the assigned reviewer?

I added him.

By the way, can you please also change the documentation about tables in docs/Tables.md ?

Lucas-C commented on fpdf/html.py:
```diff
@@ -487,6 +487,7 @@ def handle_starttag(self, tag, attrs):
                 # pylint: disable=protected-access
                 self.table._borders_layout = TableBordersLayout.NONE
                 self.table._first_row_as_headings = False
```
This line could maybe be removed?

Lucas-C commented on test/table/test_table.py:
```diff
@@ -480,3 +480,18 @@ def test_table_with_no_headers_nor_horizontal_lines(tmp_path):  # discussion 924
     )
+
+
+def test_table_with_multiple_headings_and_pagebreak(tmp_path):
```
Could you please also add unit tests for the cases you handled when both `first_row_as_headings` & `num_heading_rows` are provided?

Lucas-C commented on fpdf/table.py:
```diff
@@ -359,7 +380,7 @@ def _render_table_cell(
         if not isinstance(text_align, (Align, str)):
             text_align = text_align[j]
-        if i == 0 and self._first_row_as_headings:
+        if i < self._num_heading_rows:
```
👍 Well done

Lucas-C commented on fpdf/table.py:
```diff
@@ -282,7 +303,7 @@ def get_cell_border(self, i, j):
             if j == columns_count - 1:
                 border.remove("R")
         if self._borders_layout == TableBordersLayout.NO_HORIZONTAL_LINES:
```
Could you please use this layout in a unit test with `num_heading_rows > 1` , so that we have a non-regression test checking the expected behaviour in that case

Lucas-C commented:
@allcontributors please add SandraFer for code

Lucas-C commented on test/table/test_table.py:
```diff
@@ -493,7 +493,51 @@ def header(self):
+    pdf = FPDF()
+    pdf.set_font("Times", size=12)
+    pdf.add_page()
+    with pytest.raises(ValueError):
+        with pdf.table(TABLE_DATA, first_row_as_headings=True, num_heading_rows=0):
+            pass
+    with pytest.raises(ValueError):
+        with pdf.table(TABLE_DATA, first_row_as_headings=False, num_heading_rows=2):
+            pass
```
👍 Nice, perfect!

Lucas-C commented:
Thank you @SandraFer for the improvments you made following my feddbacks.

This seems almost ready to be merged! 🙂

I have just made a couple of tests using `borders_layout="MINIMAL"` & `borders_layout="SINGLE_TOP_LINE"`,
and the results are not really what I expected...

I think there are missing horizontal lines for those cases.
Could you please add a couple of extra unit tests for those cases,
and see how to correct `Table.get_cell_border()` in order to better handle those situations, please?

Lucas-C commented:
Btw, I added the `hacktoberfest-accepted` label to this PR in case you want to take part to https://hacktoberfest.com

SandraFer commented:
@Lucas-C thanks for adding the hacktoberfest label! I've never participated, but I just singed up for it!

I hope you agree with the new testcases! I named them such that they run as well when invoking `pytest -k table_with_single_top_line_layout` or `pytest -k table_with_minimal_layout`

Lucas-C commented:
Thank you!

Just one last nitpick: with `borders_layout="SINGLE_TOP_LINE"` I wonder if a single top line would be preferable when there are 2 heading rows...
Currently there are 2 of them: https://github.com/SandraFer/fpdf2/blob/issue-938/test/table/table_with_single_top_line_layout_and_multiple_headings.pdf

What do you think of this specific case @SandraFer, @andersonhc, @gmischler?

andersonhc commented:
> What do you think of this specific case

In my opinion single_top_line should be 1 line under the last heading row - no line between the heading rows

gmischler commented:
I think it would be helpful if the multi-header examples would more closely reflect the typical use case, with column groups like in the example in the issue #938. That would make it easier to decide how many horizontal lines are actually needed to understand the grouping.

SandraFer commented:
Thanks for the feedback @andersonhc , @gmischler 

I agree with @gmischler that we need column groups to decide what the different table layouts should look like. So, let's go through the different layouts with a better example:

1. MINIMAL
The choice here is whether we want to draw horizontal lines between the heading rows.

Choice 1: with horizontal lines between all heading rows
<img width="920" alt="image" src="https://github.com/py-pdf/fpdf2/assets/50599578/11917a28-5795-486a-83da-b5738bbb4f94">
Choice 2: horizontal lines only between last heading row and first default row
<img width="920" alt="image" src="https://github.com/py-pdf/fpdf2/assets/50599578/696bd960-50bc-4beb-b9ae-59c827aec5e5">
Frankly, I prefer the first one. The second choice looks a bit weird with the "floating" top cells `Fruits` and `Dairy`. 

(Also, there seems to be a bug when the top right cell has `colspan>1`, as the right border is drawn. Should we open a separate ticket for this?)

The same decision as for the minimal layout should be applied to "NO HORIZONTAL LINES". (We had already discussed to keep the horizontal lines between heading rows in https://github.com/py-pdf/fpdf2/issues/938#issuecomment-1744911831, but nothing is set in stone 😉 )

2. SINGLE TOP LINE
Same choice as for minimal case: do we draw all horizontal lines between heading rows?

Choice 1: draw them
<img width="920" alt="image" src="https://github.com/py-pdf/fpdf2/assets/50599578/18894c64-318d-45de-88a4-9ca0c29b311d">

Choice 2: do not draw them
<img width="920" alt="image" src="https://github.com/py-pdf/fpdf2/assets/50599578/5318e2aa-06da-4d8a-a57a-c9af9d357653">

I have a slight preference for choice 1, but choice 2 is also fine for me. In contrast to the minimal layout, there are no vertical lines here, so choice 2 does not look weird.

Let me know which variant you all prefer and I'll push it together with the improved test cases using the better table.

Lucas-C commented:
>  Frankly, I prefer the first one. The second choice looks a bit weird with the "floating" top cells `Fruits` and `Dairy`.

I agree.

> Also, there seems to be a bug when the top right cell has colspan>1, as the right border is drawn. Should we open a separate ticket for this?

Yes please!

> The same decision as for the minimal layout should be applied to "NO HORIZONTAL LINES". (We had already discussed to keep the horizontal lines between heading rows in https://github.com/py-pdf/fpdf2/issues/938#issuecomment-1744911831, but nothing is set in stone 😉 )

I think we miss a unit test with `borders_layout="NO_HORIZONTAL_LINES"` and `first_row_as_headings=True` to properly cover this case. Personnally, I was expecting zero horizontal line with `borders_layout="NO_HORIZONTAL_LINES"`, even below the headings.

> Let me know which variant you all prefer and I'll push it together with the improved test cases using the better table.

I suggest that we define a class property `Table.DRAW_ALL_HEADING_LINES = True`.
By default, when `num_heading_rows > 0` and the `layout` induces to draw horizontal lines, an horizontal line would be drawn below each heading row. But if `DRAW_ALL_HEADING_LINES is False`, then only a single line would be drawn after all heading rows.

Using such class properties as a way to provide fine-tuning configuration settings is something we already did in several places:
* `FPDF.HTML2FPDF_CLASS`
* `FPDF.MARKDOWN_*`
* `HTML2FPDF.HTML_UNCLOSED_TAGS`

I think it can sometimes be a good approach to allow users really fine control over settings
that 95% of users usually do not care to alter, because the default values are fine in most situations.

What do you think of this plan?

gmischler commented:
> (Also, there seems to be a bug when the top right cell has `colspan>1`, as the right border is drawn. Should we open a separate ticket for this?)

You can. But nobody will think worse of you if you manage to just fix it while you're fiddling with the code anyway... :wink:

Other than that, I agree that with colspan, lines under all headings make it look less confusing.

SandraFer commented:
> So I say that we just choose to draw a line below every heading row when `num_heading_rows > 0` and the `layout` induces to draw them.
> 
> What do you think of this plan?

That sound very reasonable to me, let's do this!

Lucas-C commented on test/table/test_table.py:
```diff
@@ -493,7 +502,132 @@ def header(self):
+    )
+
+
+def test_table_with_no_horizontal_lines_layout(tmp_path):
```
👍 Thank you!

Lucas-C commented on fpdf/table.py:
```diff
@@ -260,7 +281,17 @@ def get_cell_border(self, i, j):
         if self._borders_layout == TableBordersLayout.NONE:
             return 0
-        columns_count = max(row.cols_count for row in self.rows)
+        # In order to figure out whether a cell is the rightmost in its row, we
+        # need to access `colspan` of the corresponding `Cell` instance.
+        # That is because the `j` argument received from `_render_table_cell`
+        # is incremented with the colspan from cell to cell
+        columns_count = len(self.rows[i].cells)
+        colidx = 0
+        column_indexes = [colidx]
+        for jj in range(columns_count - 1):
+            colidx += self.rows[i].cells[jj].colspan
+            column_indexes.append(colidx)
+        rightmost_column = j == column_indexes[-1]
```
```suggestion
        is_rightmost_column = j == self.rows[i].column_indices[-1]
```
I think this computation would be better placed inside a dedicated method or `@property` of the `Row` class.

SandraFer replied:
Nice suggestion, makes the code a lot cleaner 😄

SandraFer commented:
Finally had some time to work on this 😄 

I added the improved test cases using the table with `colspan`, managed to fix the bug regarding the rightmost border and also added a unit test for
> unit test with borders_layout="NO_HORIZONTAL_LINES" and first_row_as_headings=True 

As per https://py-pdf.github.io/fpdf2/fpdf/enums.html#fpdf.enums.TableBordersLayout, it was on purpose to keep the horizontal line between headings and default row, so I did not touch that part

Lucas-C commented:
> As per [py-pdf.github.io/fpdf2/fpdf/enums.html#fpdf.enums.TableBordersLayout](https://py-pdf.github.io/fpdf2/fpdf/enums.html#fpdf.enums.TableBordersLayout), it was on purpose to keep the horizontal line between headings and default row, so I did not touch that part

You did right, I think I mixed things up when suggesting to remove them 🙂

I reviewed your latest changes it all seems great!
I just suggested a minor refactoring

Lucas-C commented:
Thank you @SandraFer for your great work on this!

I merged this MR, it should be included in a new release published this week.

SandraFer commented:
Thanks @Lucas-C for the review!
